Support Chinese punctuations

yizhuan opened the pull request:
Added basic punctuation support to make the reading more natural. Non-stop reading sounds weird.

I also removed jieba. I found Jieba is not useful and it made the reading pause at the wrong positions. Sentence separation is now fully decided by punctuation.

hexgrad commented on misaki/zh.py:
```diff
@@ -1,10 +1,25 @@
 
 class ZHG2P:
+
+    punctuations = r'[，、：；。！？]'
```
Should we also handle the ASCII punctuation marks like `,:;.!?`, and normalize with something like `unicodedata.normalize('NFKC', text)`?

yizhuan replied:
let me unicodedata.normalize. didn't know about it :-)

hexgrad commented on misaki/zh.py:
```diff
@@ -1,10 +1,25 @@
+
+    punctuations = r'[，、：；。！？]'
+    punctuation_replacements = {
+            '，': ' | ',
+            '、': ' | ',
+            '：': ' | ',
+            '；': ' ‖ ‖ ',
+            '。': ' ‖ ‖ ‖ ',
+            '！': ' ‖ ‖ ‖ ', 
+            '？': ' ‖ ‖ ‖ ', 
```
Unfortunately this is a breaking change for `kokoro` since those vertical bars are not in the active vocabulary: https://huggingface.co/hexgrad/Kokoro-82M/blob/main/config.json#L34

The vocabulary can be changed if the model is retrained, but separately, for modeling reasons we would probably rather have single tokens per punctuation rather than expanding the tokens like this.

yizhuan replied:
do you mean, not "‖ ‖ ‖" but can be adjusted to e.g. one bar " ‖ ". Let me test and get back.

If the bars are not in the vocabulary, then Chinese audio problem is primarily caused by jieba.

yizhuan replied:
I just had a quick test. The bars do not have real effect (confirms they are not in vocabulary). What made the difference is the spaces with the bar I added. I found that just replacing all punctuations with a "space" has the same effect as the "bars" in effect, greatly improves the audio quality. space characters have a pause (stop) effect.

hexgrad reviewed (commented):
Breaking vocabulary changes, see comments.

> I found Jieba is not useful and it made the reading pause at the wrong positions.

Can you tell if Jieba at fault or the model (Kokoro), or both?
If you're reading the text, and it looks cut in the wrong place, it could be Jieba's fault.
But, if Jieba is cutting in the right place, and Kokoro is pausing too long or too little, then it could be the model's fault (likely bad training data).

yizhuan commented:
> Breaking vocabulary changes, see comments.
> 
> > I found Jieba is not useful and it made the reading pause at the wrong positions.
> 
> Can you tell if Jieba at fault or the model (Kokoro), or both? If you're reading the text, and it looks cut in the wrong place, it could be Jieba's fault. But, if Jieba is cutting in the right place, and Kokoro is pausing too long or too little, then it could be the model's fault (likely bad training data).

"Jieba" in Chinese means "stammer" or "stutter".  Jieba's primary purpose is not speech synthetics. It is more for search or data collection, as I understand. jieba makes kokoro speech synthetics sound weird. The problem is more about it breaks and pauses at wrong place. Sometimes a "Chinese word" is broken in the middle, which is obviously wrong. I observed that even with my very limited tests. 

I am not sure whether there is any problem with the model on this issue.

In general, after I removed jieba and added punctuation support, the result is near "useful" quality (pretty good, but still not as good as English audio). Without these changes, the produced audio is pretty much useless.

I will check other issues you raised.

hexgrad commented:
> "Jieba" in Chinese means "stammer" or "stutter". Jieba's primary purpose is not speech synthetics. It is more for search or data collection, as I understand. jieba makes kokoro speech synthetics sound weird. The problem is more about it breaks and pauses at wrong place. Sometimes a "Chinese word" is broken in the middle, which is obviously wrong. I observed that even with my very limited tests.

Thanks, this is very useful feedback. There is a chance I might get my hands on substantially better Chinese data soon (no promises), and if so, the next Chinese-capable model will definitely warrant revisiting or removing jieba from the equation.

> I am not sure whether there is any problem with the model on this issue.
> 
> In general, after I removed jieba and added punctuation support, the result is near "useful" quality (pretty good, but still not as good as English audio). Without these changes, the produced audio in Chinese is pretty much useless.

The thing is, the model was already trained on `jieba` as well. Even if removing `jieba` makes it sound better at inference, the model is still handicapped from its peak potential. (Models can always be retrained, however.)

hexgrad commented:
@yizhuan The root cause of Chinese punctuation not working may be the failure to convert full-width punctuation back to ASCII, see https://github.com/hexgrad/kokoro/issues/59#issuecomment-2644244673
